Corgibytes.Freshli.Lib.Test/Unit/Util/DateTimeExtensionsTest.cs

```diff
@@ -1,49 +1,240 @@
+      new[] {2020, 1, 6, 12, 0, 0, 0, 0},
+      new[] {2020, 01, 01, 00, 00, 00, 000, 0000},
+      0
+    )]
+    [InlineData(
+      nameof(DateTimeExtensions.ToStartOfMonth),
+      new[] {2020, 1, 6, 23, 59, 59, 999, 9998},
+      new[] {2020, 01, 01, 00, 00, 00, 000, 0000},
+      0
+    )]
+    [InlineData(
+      nameof(DateTimeExtensions.ToStartOfMonth),
+      new[] {2020, 1, 6, 23, 59, 59, 999, 9999},
+      new[] {2020, 01, 01, 00, 00, 00, 000, 0000},
+      0
+    )]
+    [InlineData(
+      nameof(DateTimeExtensions.ToStartOfMonth),
+      new[] {2020, 1, 7, 0, 0, 0, 0, 0},
+      new[] {2020, 01, 01, 00, 00, 00, 000, 0000},
+      0
+    )]
+    [InlineData(
+      nameof(DateTimeExtensions.ToStartOfMonth),
+      new[] {2020, 1, 7, 0, 0, 0, 0, 1},
+      new[] {2020, 01, 01, 00, 00, 00, 000, 0000},
+      0
+    )]
+    [InlineData(
+      nameof(DateTimeExtensions.ToStartOfMonth),
+      new[] {2020, 2, 7, 0, 0, 0, 0, 0},
+      new[] {2020, 01, 01, 00, 00, 00, 000, 0000},
+      1
+    )]
+    [InlineData(
+      nameof(DateTimeExtensions.ToStartOfMonth),
+      new[] {2020, 2, 7, 0, 0, 0, 0, 1},
+      new[] {2020, 01, 01, 00, 00, 00, 000, 0000},
+      1
+    )]
+    public void EndOfDayValuesCompareCorrectly(
+      string methodName,
+      int[] dateArguments,
+      int[] targetDateArguments,
       int comparison
     ) {
-      var inputDate = new DateTime(
-        year,
-        month,
-        day,
-        hour,
-        minute,
-        second,
-        millisecond
-      ).AddTicks(tick);
-
-      var inputDateEoD = inputDate.ToEndOfDay();
-
-      Assert.Equal(
+      AssertDateExtensionMethod(
+        dateArguments,
+        targetDateArguments,
         comparison,
-        inputDateEoD.CompareTo(_baseDateTime)
+        date => InvokeExtensionMethod(date, methodName)
       );
     }
+
+    private static DateTimeOffset InvokeExtensionMethod(
+      DateTimeOffset date,
+      string methodName
+    ) {
+      return (DateTimeOffset) typeof(DateTimeExtensions).GetMethod(methodName).
+        Invoke(date, new object[] {date});
+    }
+
+    private static DateTimeOffset BuildDateTimeOffset(int[] dateArguments)
+    {
+      var inputDate = new DateTimeOffset(
+        dateArguments[0],
+        dateArguments[1],
+        dateArguments[2],
+        dateArguments[3],
+        dateArguments[4],
+        dateArguments[5],
+        dateArguments[6],
+        TimeSpan.Zero
+      ).AddTicks(dateArguments[7]);
+      return inputDate;
+    }
+
+    private void AssertDateExtensionMethod(
+      int[] dateArguments,
+      int[] targetDateArguments,
+      int comparison,
+      Func<DateTimeOffset, DateTimeOffset> extensionMethod
+    ) {
+      var inputDate = BuildDateTimeOffset(dateArguments);
+      var targetDate = BuildDateTimeOffset(targetDateArguments);
+
+      var transformedDate = extensionMethod(inputDate);
+
+      Assert.Equal(comparison, transformedDate.CompareTo(targetDate));
+    }
```

Comment:
Same thoughts for these DateTimeOffset extension tests as the Repository tests.  What do you think about having tests for the each of the extensions that call common test methods?

Reply:
I think that something similar to my suggestion for the IRepository testers could be used here as well.

```csharp
public class TestDateTimeOffsetExtensions
{
    public class ToEndOfDay: DateModificationExtensionMethodTestFixture
    {
        protected override DateTimeOffset InvokeMethod(DateTimeOffset value) => value.ToEndOfDay();

        protected override TheoryData<IList<int>, IList<int>, int> DataForTestingExtensionMethod => new()
        {
            { 
                 new[] {2020, 1, 5, 23, 59, 59, 999, 9998},
                 new[] {2020, 1, 6, 23, 59, 59, 999, 9999},
                 -1
            },
            // additional test data snipped
        }
    }

    public class ToStartOfDay: DateModificationExtensionMethodTestFixture
    {
        protected override DateTimeOffset InvokeMethod(DateTimeOffset value) => value.ToStartOfDay();

        // contents snipped
    }

}

Would that work?
```

Reply:
I like that idea.  That way everyone can easily find the test file.  They can also add tests specific to that extension if needed.